db/src/db.rs

```diff
@@ -0,0 +1,532 @@
+        r#"CREATE VIEW all_datoms AS
+             SELECT e, a, v, tx, value_type_tag, index_avet, index_vaet, index_fulltext, unique_value
+               FROM datoms
+               WHERE index_fulltext IS 0
+             UNION ALL
+             SELECT e, a, v, tx, value_type_tag, index_avet, index_vaet, index_fulltext, unique_value
+               FROM fulltext_datoms"#,
+
+        // Materialized views of the schema.
+        r#"CREATE TABLE idents (ident TEXT NOT NULL PRIMARY KEY, entid INTEGER UNIQUE NOT NULL)"#,
+        r#"CREATE TABLE schema (ident TEXT NOT NULL, attr TEXT NOT NULL, value BLOB NOT NULL, value_type_tag SMALLINT NOT NULL,
+             FOREIGN KEY (ident) REFERENCES idents (ident))"#,
+        r#"CREATE INDEX idx_schema_unique ON schema (ident, attr, value, value_type_tag)"#,
+        r#"CREATE TABLE parts (part TEXT NOT NULL PRIMARY KEY, start INTEGER NOT NULL, idx INTEGER NOT NULL)"#,
+        ]
+    };
+}
+
+/// Set the SQLite user version.
+///
+/// Mentat manages its own SQL schema version using the user version.  See the [SQLite
+/// documentation](https://www.sqlite.org/pragma.html#pragma_user_version).
+fn set_user_version(conn: &rusqlite::Connection, version: i32) -> Result<()> {
+    conn.execute(&format!("PRAGMA user_version = {}", version), &[])
+        .chain_err(|| "Could not set_user_version")
+        .map(|_| ())
+}
+
+/// Get the SQLite user version.
+///
+/// Mentat manages its own SQL schema version using the user version.  See the [SQLite
+/// documentation](https://www.sqlite.org/pragma.html#pragma_user_version).
+fn get_user_version(conn: &rusqlite::Connection) -> Result<i32> {
+    conn.query_row("PRAGMA user_version", &[], |row| {
+        row.get(0)
+    })
+        .chain_err(|| "Could not get_user_version")
+}
+
+// TODO: rename "SQL" functions to align with "datoms" functions.
+pub fn create_current_version(conn: &mut rusqlite::Connection) -> Result<i32> {
+    let tx = conn.transaction()?;
+
+    for statement in (&V2_STATEMENTS).iter() {
+        tx.execute(statement, &[])?;
+    }
+
+    let bootstrap_partition_map = bootstrap::bootstrap_partition_map();
+    // TODO: think more carefully about allocating new parts and bitmasking part ranges.
+    // TODO: install these using bootstrap assertions.  It's tricky because the part ranges are implicit.
+    // TODO: one insert, chunk into 999/3 sections, for safety.
+    for (part, partition) in bootstrap_partition_map.iter() {
+        // TODO: Convert "keyword" part to SQL using Value conversion.
+        tx.execute("INSERT INTO parts VALUES (?, ?, ?)", &[part, &partition.start, &partition.index])?;
+    }
+
+    let bootstrap_db = DB::new(bootstrap_partition_map, bootstrap::bootstrap_schema());
+    bootstrap_db.transact_internal(&tx, &bootstrap::bootstrap_entities()[..])?;
+
+    set_user_version(&tx, CURRENT_VERSION)?;
+    let user_version = get_user_version(&tx)?;
+
+    // TODO: use the drop semantics to do this automagically?
```

Comment:
I wouldn't recommend this (i.e., I'd keep it explicit as written). By default dropping `tx` would attempt to perform a rollback. You could change that with `set_drop_behavior`, but then you don't get to see if the commit succeeded or failed.